File: 2816-double-a-number-represented-as-a-linked-list/2816-double-a-number-represented-as-a-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def doubleIt(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # Step 1: Convert linked list to string number
        num_str = ""
        temp = head
        while temp:
            num_str += str(temp.val)
            temp = temp.next
        
        # Step 2: Double the number manually
        carry = 0
        doubled_digits = []
        for digit in reversed(num_str):
            prod = int(digit) * 2 + carry
            doubled_digits.append(str(prod % 10))
            carry = prod // 10
        if carry:
            doubled_digits.append(str(carry))
        
        # Step 3: Rebuild the linked list from doubled number
        doubled_digits.reverse()
        new_head = ListNode(int(doubled_digits[0]))
        curr = new_head
        for d in doubled_digits[1:]:
            curr.next = ListNode(int(d))
            curr = curr.next
        
        return new_head

File: 2816-double-a-number-represented-as-a-linked-list/2816-double-a-number-represented-as-a-linked-list.java
class Solution {
    public ListNode doubleIt(ListNode head) {
        StringBuilder sb = new StringBuilder();
        ListNode temp = head;
        while (temp != null) {
            sb.append(temp.val);
            temp = temp.next;
        }

        String num = sb.toString();
        StringBuilder doubled = new StringBuilder();
        int carry = 0;

        for (int i = num.length() - 1; i >= 0; i--) {
            int digit = num.charAt(i) - '0';
            int prod = digit * 2 + carry;
            doubled.append(prod % 10);
            carry = prod / 10;
        }
        if (carry > 0) {
            doubled.append(carry);
        }
        doubled.reverse();

        ListNode newHead = new ListNode(doubled.charAt(0) - '0');
        ListNode current = newHead;
        for (int i = 1; i < doubled.length(); i++) {
            current.next = new ListNode(doubled.charAt(i) - '0');
            current = current.next;
        }

        return newHead;
    }
}
